app.rb

```diff
@@ -18,194 +18,288 @@
+        files: Directory.new(path).files,
+      }
     end
 
-    redirect to("/")
-  end
+    def normalize(data)
+      case data
+      when Array
+        data.map &method(:normalize)
+      when Hash
+        if data.keys.include?('___checkbox_unchecked')
+          data.keys.include?('___checkbox_checked')
+        elsif data.keys.all? { |k| k =~ /\A\d+\Z/ }
+          data.to_a.sort_by(&:first).map(&:last).map &method(:normalize)
+        else
+          data.merge(data) { |k,v| normalize(v) }
+        end
+      else
+        data
+      end
+    end
 
-protected
+    helpers do
+      def render_yml(value)
+        case value
+        when Hash
+          value.map { |k, v| render_value(k, v, "data[#{k}]") }.join
+        when Array
+          value.each_with_index.map { |v, i| render_value(nil, v, "data[#{i}]") }.join
+        else
+          raise TypeError, 'YAML root must be a Hash or Array'
+        end
+      end
 
-  def detect_content_or_frontmatter(path)
-    has_frontmatter = false
-    File.foreach(path).with_index do |line, idx|
-      line.chomp!
-      if line == '---' && idx == 0
-        has_frontmatter = true
-        next
+      def render_value(key, value, name)
+        case value
+        when Hash
+          "<fieldset>#{"<legend>#{key}</legend>" if key}#{
+          value.map do |key, value|
+            "#{render_value(key, value, "#{name}[#{key}]")}"
+          end.join
+          }</fieldset>"
+        when Array
+          "<fieldset><legend>#{key}</legend><ol>#{
+            value.each_with_index.map { |v, i| "<li>#{render_value(nil, v, "#{name}[#{i}]")}</li>" }.join
+          }</ol></fieldset>"
+        else
+          render_input(key, value, name)
+        end
       end
 
-      if has_frontmatter && line == '---'
-        return [true, has_frontmatter]
+      def render_input(key, value, name)
+        input = if [true, false].include? value
+          "
+            <input type='hidden' name='#{name}[___checkbox_unchecked]' value='' />
+            <input type='checkbox' name='#{name}[___checkbox_checked]' #{' checked="checked"' if value} />
+          "
+        elsif value.is_a?(String) && value.length > 50
+          "<textarea name='#{name}'>#{value}</textarea>"
+        else
+          "<input type='text' name='#{name}' value='#{value}' />"
+        end
+
+        if key
+          "<label>#{key} #{input}</label>"
+        else
+          input
+        end
       end
     end
-
-    [!has_frontmatter, has_frontmatter]
   end
 
-  def stream_content(path)
-    has_frontmatter = false
-    doc_start = 0
-    File.foreach(path).with_index do |line, idx|
-      line.chomp!
+  class Directory
+    def initialize(path)
+      @path = path
+    end
 
-      if doc_start < 2
-        has_frontmatter = true if line == '---' && idx == 0
-        doc_start += 1 if line == '---'
-        next if has_frontmatter
+    def files
+      files = Dir.glob("#{path}/*").map { |file| Tint::File.new(file) }
```

Comment:
With `Dir` operations in general, we need to be careful that we're sanitizing `path` -- probably before it ever gets this far.  So that it does not contain `..` or similar.  I don't know how we're going to sandbox this stuff long-term, but we'd like to avoid people editing random files on the host system, I'm sure :)